### PR TITLE
docker compose with multi stage

### PR DESCRIPTION
**Pull request by hossainahmed18**
Fixes #

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [ ] New feature
- [ ] Refactor
- [x] Infrastructure
- [ ] Other (please specify)

## What was done

docker compose files are introduced. Firstly, files are changed to build a docker image and run react app on container. Then in next step docker compose files are changed for multi stage deployment. React app will be deployed on nginx server

## How to test

- [x] Needs regression testing (add QA needed label)

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

## Reminder list

- Determine whether REST API Documentation should be updated
- Make an assessment on whether unit tests should be updated and/or added
- Ensure follow up tasks are linked to this PR

QA - if regression testing was requested:

- [ ] QA has verified

### CODE REVIEW

File: junayed-task/todo-docker/Dockerfile.prod [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#This will build the react app
+RUN npm run build
+
+
+#pulling the base image for ngnix
+FROM nginx:alpine
+#copying files from container work directory to ngnix public folder
+COPY --from=node app/build /usr/share/nginx/html
+Expose 4200
```
Comment (musfique-m) [inline junayed-task/todo-docker/Dockerfile.prod:24]: Why are port 4200 exposed here?
nginx listens to port 80 by default. This EXPOSE instruction should indicate what ports are we gonna get our output from this Dockerfile.

File: junayed-task/todo-docker/Dockerfile.prod [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#pulling the base image from docker hub
+FROM node:alpine as node
+
+#creating a work directory in docker container
+WORKDIR /app
+
+#This will prevent unncessarily runing "npm install" if there is no change in pachage.json file
+COPY package.json .
```
Comment (musfique-m) [inline junayed-task/todo-docker/Dockerfile.prod:8]: Please also copy `package-lock.json` to keep track of the exact version of every package that is installed.
Also use wild card to reduce the number of layers.

Also please remove comments from prod files.

**Review by musfique-m — CHANGES_REQUESTED** [review]

File: junayed-task/todo-docker/Dockerfile.prod (hunk 1 above)
Comment (hossainahmed18) [reply junayed-task/todo-docker/Dockerfile.prod:24]: In docker-compose file I mapped port 80 to 4200. That's why I exposed it to 4200. 
image

File: junayed-task/todo-docker/Dockerfile.prod (hunk 2 above)
Comment (hossainahmed18) [reply junayed-task/todo-docker/Dockerfile.prod]: resolved with commit : https://github.com/CraftsmenLtd/coe-batch-q322/pull/296/commits/ac7c4bacd21678cf74b4a54489cdb73a54b2228f

File: junayed-task/todo-docker/Dockerfile.prod [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+#This will build the react app
+RUN npm run build
+
+
+#pulling the base image for ngnix
+FROM nginx:alpine
+#copying files from container work directory to ngnix public folder
+COPY --from=node app/build /usr/share/nginx/html
+Expose 4200
```
Comment (musfique-m) [inline junayed-task/todo-docker/Dockerfile.prod:24]: This expose indicates the port that we are going to listen to in the container. The container is run from an image that is built with the help of this Dockerfile. So we should always EXPOSE the port here that we are going to use. In this case it is 80, not 4200. In our docker-compose we are just mapping port 4200 of the host os to the port 80 of this container. We can map that to many ports as we want. But our container built from our image will only be accessible by the port 80 of the container.

**Review by musfique-m — CHANGES_REQUESTED** [review]
Please look at the comment about port expose.

Also remove comment from prod files.

File: junayed-task/todo-docker/Dockerfile.prod (hunk 1 above)
Comment (musfique-m) [reply junayed-task/todo-docker/Dockerfile.prod]: Please look at the comment.

File: junayed-task/todo-docker/Dockerfile.prod (hunk 3 above)
Comment (hossainahmed18) [reply junayed-task/todo-docker/Dockerfile.prod]: resolved with last commits

**Review by musfique-m — CHANGES_REQUESTED** [review]

File: junayed-task/todo-docker/Dockerfile.prod [hunk 4]
```diff
@@ -0,0 +1,9 @@
+FROM node:alpine as node
+WORKDIR /app
+COPY package*.json .
+RUN npm install
+COPY . ./
+RUN npm run build
+FROM nginx:alpine
+COPY --from=node app/build /usr/share/nginx/html
+Expose 80
```
[the remark this answers is not in the document]
Comment (hossainahmed18) [reply junayed-task/todo-docker/Dockerfile.prod]: resolved with commit : https://github.com/CraftsmenLtd/coe-batch-q322/pull/296/commits/9ec4d96fbc62067fb935f4001670feab63f2e85b

File: junayed-task/todo-docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+#This will copy rest of the file
+COPY . ./
+
+
+#This will run in CMD and it will react app in docker container
+CMD npm start
+
+Expose 3000
```
[the remark this answers is not in the document]
Comment (hossainahmed18) [reply junayed-task/todo-docker/Dockerfile]: resolved with commit: https://github.com/CraftsmenLtd/coe-batch-q322/pull/296/commits/9ec4d96fbc62067fb935f4001670feab63f2e85b

**Review by musfique-m — APPROVED** [review]
Looks good.

**Pull request merged**